### ISSUE TITLE
Release/Branch Strategy

### ISSUE THREAD

**Issue by hahn-kev** [sillsdev/languageforge-lexbox#59]
Language Forge uses the following release strategy:
- 2 branch strategy, develop and main.
- quickly able to deploy a hotfix by committing directly to main
- low ceremony.
- if develop isn't in a good state (maybe a long running task), we can't test and deploy to prod, some completed work can't be shipped which delays testing.
- we can't have both a dev and staging environment even if we wanted to since they are the same.
- the only way to test a PR is by checking it out and testing locally.

We'd like to decide if we want to continue using that release strategy in this new repository, or if we want to make some adjustments.

I'd like to propose (and implement myself) a slightly different way of doing the deployment.

Github actions now supports manual approvals with environments. This means that before an action that happens against an environment we can require manual approval from someone on the team, they would also be notified of this via email (and maybe slack?). [A blog about it](https://cloudlumberjack.com/posts/github-actions-approvals/) and the [github docs](https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments).

What I would like to try out is to build and deploy main to staging on push, and that would kick off a manual approval to deploy to prod (theoretically after testing on staging). This would then free up develop as it's not always getting pushed to staging, and if we want to deploy some changes we can merge only some of the changes in develop to main, not latest (which might not be ready to be deployed).

In the future I'd also like to setup a develop environment that we can deploy builds from pr's and other branches. Chris mentioned he's liked having an environment like that on other projects and I think that would be helpful for PR testing as it's less disruptive to the dev since they don't have to test locally.

@sillsdev/language-forge leave your comments below.

**Comment by megahirt** [comment]
@hahn-kev thanks for this.  We talked about this in the office, but I'll summarize my thoughts here.  I am definitely open to this release/branch strategy.

@longrunningprocess and I spent a long time over the past two years refining our release branch strategy for LF and I'm happy with where things are there.  When we started down the journey of refactoring LF CI/CD, I don't think the Github environments and approval feature was available - looks like it didn't land as a beta [until the very end of 2020](https://github.blog/changelog/2020-12-15-github-actions-environments-environment-protection-rules-and-environment-secrets-beta/).

Some things I like about this new strategy:
- we don't have to re-build the image for production.  We just ship the image we tested in staging.  In practice I've never seen an issue in our current "re-build a production image" CI since the code should be identical.  We have lock files for our dependencies...not sure how that could go wrong, but I've seen a few GHA hiccups over the years.  Saving the rebuild time and simplifying CI sounds like a win in this case.
- it seems likely that our deployment scripts will be simpler if engineered this way - i'm a big fan of simplicity!

Some things about this that are new to me, and I think are worth proving out:
- our release branch (main/master) would deploy to both staging and prod.  This seems logical to me: the image goes on a journey through staging, gets tested and approved, and lands in production.
- if the develop branch isn't shippable (long running epic/story with multiple PRs), doing selective merges and/or cherry-picking commits could work to help us ship faster, however this strategy isn't full proof IMO and sometimes depends on how extensive changes are.  It's worth exploring.

Questions:
- do we still build a develop image on every commit?  I understand it wouldn't necessarily automatically deploy

In summary, we would have three deployments:
- dev (never automatically deployed, but can be manually deployed with a given PR or develop image using workflow dispatch)
- staging
- prod

**Comment by longrunningprocess** [comment]
the fewer branches the better in my mind, I'm interested the experiment.
